ibtk/src/lagrangian/FEDataManager.cpp

```diff
@@ -842,19 +835,19 @@ FEDataManager::spread(const int f_data_idx,
 
     if (use_nodal_quadrature)
     {
-        // Store a copy of the F vector.
-        std::unique_ptr<NumericVector<double> > F_vec_bak = F_vec.clone();
-        *F_vec_bak = F_vec;
-
         // Multiply by the nodal volume fractions (to convert densities into
         // values).
-        NumericVector<double>& F_dX_vec = F_vec;
-        const NumericVector<double>& dX_vec = *buildDiagonalL2MassMatrix(system_name);
-        F_dX_vec.pointwise_mult(F_vec, dX_vec);
-        F_dX_vec.close();
+        std::unique_ptr<NumericVector<double> > F_dX_vec = F_vec.clone();
+        std::unique_ptr<NumericVector<double> > dX_vec = F_dX_vec->clone();
+        *dX_vec = *buildDiagonalL2MassMatrix(system_name);
+        F_dX_vec->pointwise_mult(F_vec, *dX_vec);
 
         // Extract local form vectors.
-        auto F_dX_petsc_vec = static_cast<PetscVector<double>*>(&F_dX_vec);
+        auto F_dX_petsc_vec = static_cast<PetscVector<double>*>(F_dX_vec.get());
+        int ierr = VecGhostUpdateBegin(F_dX_petsc_vec->vec(), INSERT_VALUES, SCATTER_FORWARD);
+        IBTK_CHKERRQ(ierr);
+        ierr = VecGhostUpdateEnd(F_dX_petsc_vec->vec(), INSERT_VALUES, SCATTER_FORWARD);
+        IBTK_CHKERRQ(ierr);
```

Comment:
We typically don't do communication inside FEDataManager. Could this be moved somewhere else?

Reply:
Perhaps, but we would somehow need to know the layout of the `F` vector in order to setup the ghosted entries of the vector of the diagonal mass matrix entries.

An alternative could be to apply the diagonal mass matrix in a matrix-free manner.

Reply:
Anyway, if you have suggestions on where to move it / how to avoid it, I am happy to do it; but I couldn't think of a simple way to do it, which is why it is where it is now.

Reply:
I don't understand why this is necessary. Applying the diagonal mass matrix should be a purely local operation. Let me think about this a little more.

Reply:
Because we need the forces on the IB ghost nodes. Upon entry to the function, `F` has units of force density. We need to scale the forces by the mass matrix to get the nodal forces.

Reply:
(Notice that this is essentially what happens for the other code path, but it is done in a matrix-free way.)

Reply:
@drwells how is this?

Reply:
Better, but we can improve it more. Given our other problems with multiple parts I would like to completely avoid communication in this class.

Reply:
Yes I got rid of this --- this is out of date.